Helga Shiryaeva #4

VolhaAnishchanka opened the pull request.

VolhaAnishchanka committed:
Create IslandsCounter.java

VolhaAnishchanka committed:
Create IslandsCounterTest.java

brzozan commented on HelgaShiryaeva/assignment4/IslandsCounter:
```diff
@@ -0,0 +1,88 @@
+     * island of booleans with true value. The rule of exploring is that each cell is
+     * adjacent to its top, left, bottom and right neighbour.
+     * @param tiles is a grid of booleans which connected components (islands)
+     * will be counted
```
It would be good to include information about data format (that `true` indicates land) here. This would let IDE parse it out and show as a tooltip, which is very helpful, since person that wants to use this method does not have to read full description to find out what should the contents of `tiles` be.

brzozan commented on HelgaShiryaeva/assignment4/IslandsCounter:
```diff
@@ -0,0 +1,88 @@
+    private void exploreAdjacentCell(Queue<Cell> queue, Cell cell, boolean[][] tiles) {
+        int row = cell.getRow();
+        int column = cell.getColumn();
+        if(!(row < 0 || column < 0 || row >= tiles.length || column >= tiles[0].length) && tiles[row][column]) {
```
Please avoid complicating logic like that. It is quite difficult to analyse. It would be much better if it was rewritten as `(row >= 0 && column >= 0 && row < tiles.length && column < tiles.length && tiles[row][column])`.
Alternatively it could be split into two ifs - on to check coordinates, one to check if tile is `true`.

brzozan commented on HelgaShiryaeva/assignment4/IslandsCounter:
```diff
@@ -0,0 +1,88 @@
+            for (int j = 0; j < tiles[0].length; j++) {
+                if (tiles[i][j]) {
+                    queue.add(new Cell(i, j));
+                    tiles[i][j] = false;
```
This is very dangerous and should be explicitly mentioned in method comments. User of this method might not expect that the data will get modified. Imagine if someone wanted to print out map if it included more than 3 islands. They would probably write code like:

```
booelan[][] map = getMyAwesomeMap();
if (islandCounter.countIslands(map) > 3) {
  printMap(map); // what will be printed?
}
```

brzozan commented on HelgaShiryaeva/assignment4/IslandsCounter:
```diff
@@ -0,0 +1,88 @@
+     * This class represents a cell of a two-dimensional grid
+     */
+    class Cell {
+        int row;
```
Both those fields should be private and final.

brzozan commented on HelgaShiryaeva/assignment4/IslandsCounterTest:
```diff
@@ -0,0 +1,105 @@
+                {false, false, true, false},
+                {false, false, true, false}
+        };
+        IslandsCounter islandsCounter = new IslandsCounter();
```
`IslandsCounter` is stateless, so it typically would be created either directly as a field in `IslandsCounterTest` or in `setUp()` method.

brzozan reviewed (changes requested):
Very nice PR! Solution with using queue is interesting, even though a bit more difficult to analyse.

I added a few comments to the code.

VolhaAnishchanka committed:
Update IslandsCounterTest

VolhaAnishchanka committed:
Update IslandsCounter

A very good remark about unexpected modification of an input array ! 
I do agree that it can be a reason for big problems

brzozan reviewed (approved):
Thank you! LGTM.
